File: Python3/convert_sorted_array_to_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        return self.convert(nums, 0, len(nums)-1)
    
    def convert(self, nums, left, right):
        if left > right: return None
        
        mid = (left + right) // 2
        root = TreeNode(nums[mid])
        root.left = self.convert(nums, left, mid-1)
        root.right = self.convert(nums, mid+1, right)
        
        return root

File: C/convert_sorted_array_to_binary_search_tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode* convert(int* nums, int left, int right) {
    if(left > right) return NULL;
        
    int mid = (left + right) / 2;
    struct TreeNode* root = malloc(sizeof(struct TreeNode));
    root->val = *(nums+mid);
    root->left = convert(nums, left, mid-1);
    root->right = convert(nums, mid+1, right);
        
    return root;
}

struct TreeNode* sortedArrayToBST(int* nums, int numsSize) {
    return convert(nums, 0, numsSize-1);
}
